In [33]:
import pandas as pd
from sklearn.preprocessing import MinMaxScaler
from sklearn.neighbors import NearestNeighbors

In [34]:
file = pd.read_excel("Scotch_Whiskies_Data.xlsx", header=[0, 1])
df = file.copy()
df[("FIN", "REGION")] = df[("FIN", "REGION")].str.upper().str.strip()  # removing whitespaces and converting to uppercase
df[("FIN", "DISTRICT")] = df[("FIN", "DISTRICT")].str.upper().str.strip() # removing whitespaces and converting to uppercase
df

/Users/mikolajbankowski/Desktop/Data-Analyst-Task/venv-jp/lib/python3.10/site-packages/openpyxl/worksheet/header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored
  warn("""Cannot parse header or footer so it will be ignored""")
/Users/mikolajbankowski/Desktop/Data-Analyst-Task/venv-jp/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)


Unnamed: 0_level_0 color                                               \
                  NAME  wyne yellow v.pale pale p.gold gold o.gold f.gold   
0            Aberfeldy     0      1      0    0      0    0      0      0   
1             Aberlour     0      0      0    0      0    0      0      0   
2              Ardberg     0      0      0    0      0    0      0      0   
3              Ardmore     0      0      0    0      1    0      0      0   
4         Auchentoshan     0      0      0    0      1    0      0      0   
..                 ...   ...    ...    ...  ...    ...  ...    ...    ...   
104          Tobermory     0      0      0    0      0    0      0      1   
105            Tomatin     0      0      0    0      0    0      0      1   
106          Tomintoul     0      0      0    0      0    0      0      1   
107            Tormore     0      0      0    0      0    0      0      1   
108       Tullibardine     0      0      0    0      0    1      0      0   

            ...   FIN                                                       
    bronze  ... arome ling long very quick AGE SCORE     % REGION DISTRICT  
0        0  ...     0    0    0    0     0   9    69  40.0   HIGH  MIDLAND  
1        0  ...     0    1    0    0     0  12    83  43.0   HIGH     SPEY  
2        0  ...     0    0    0    0     0  10    85  40.0  ISLAY    SOUTH  
3        0  ...     0    0    0    0     0  18    66  46.0   HIGH     SPEY  
4        0  ...     0    0    0    0     0  10    85  40.0    LOW     WEST  
..     ...  ...   ...  ...  ...  ...   ...  ..   ...   ...    ...      ...  
104      0  ...     1    0    0    0     0   9    67  40.0   HIGH     MULL  
105      0  ...     0    0    0    0     0  10    75  40.0   HIGH     SPEY  
106      0  ...     0    0    1    0     0  12    76  40.0   HIGH     SPEY  
107      0  ...     0    0    0    1     0  10    76  43.0   HIGH     SPEY  
108      0  ...     1    0    0    0     0  10    76  40.0   HIGH  MIDLAND  

[109 rows x 74 columns]

### Whiskey Production and Quality by Region

In [35]:
amount_of_whiskey_per_region = df[('FIN', 'REGION')].value_counts()
average_score_per_region  = df.groupby(('FIN', 'REGION'))[[('FIN', 'SCORE')]].mean()

summary_region = pd.concat([amount_of_whiskey_per_region, average_score_per_region], axis=1)
summary_region

count  (FIN, SCORE)
(FIN, REGION)                     
HIGH              89     74.842697
LOW               12     77.333333
ISLAY              8     81.250000

### Whiskey Production and Quality by District

In [36]:
amount_of_whiskey_per_district  = df.groupby(('FIN', 'REGION'))[[('FIN', 'DISTRICT')]].value_counts()
average_score_per_district  = df.groupby([('FIN', 'REGION'), ('FIN', 'DISTRICT')])[[('FIN', 'SCORE')]].mean()
summary_district = pd.concat([amount_of_whiskey_per_district, average_score_per_district], axis=1)
summary_district

count  (FIN, SCORE)
(FIN, REGION) (FIN, DISTRICT)                     
HIGH          SPEY                57     75.403509
              EAST                 9     71.444444
              NORTH                7     77.000000
              MIDLAND              6     74.500000
              WEST                 5     68.000000
              ORKNEY               2     83.000000
              JURA                 1     71.000000
              MULL                 1     67.000000
              SKYE                 1     90.000000
ISLAY         SOUTH                4     84.750000
              LOCH                 2     78.500000
              NORTH                2     77.000000
LOW           WEST                 4     69.500000
              CAMPBEL              3     87.666667
              CENTRAL              2     71.500000
              BORDERS              1     85.000000
              EAST                 1     76.000000
              NORTHWEST            1     83.000000

#### Correlation between Age and Score of Whiskey

In [37]:
whiskey_age = df[("FIN", "AGE")]
whiskey_score = df[("FIN", "SCORE")]

correlation = whiskey_score.corr(whiskey_age)
correlation


np.float64(-0.03132004649628831)

### 1. What interesting information user can find in the dataset?

The dataset contains detailed information about various Scotch whiskies. It includes attributes such as:

- Color - Represents the visual appearance of the whisky, encoded binarily (1 for presence, 0 for absence), datatype: int
- Nose - Describes the aroma of the whisky, encoded binarily, datatype: int
- Body - Indicates the texture or mouthfeel of the whisky, encoded binarily, datatype: int
- PAL - Represents the flavor profile of the whisky, encoded binarily, datatype: int
- FIN - Describes the finish or aftertaste of the whisky, encoded binarily, datatype: int
- Region - Indicates the geographical area where the whisky is produced, datatype: string
- District - Specifies the district within the region of production, datatype: string
- AGE - The age of the whisky in years, datatype: int
- SCORE - A numerical score representing the quality or rating of the whisky, datatype: float
- % - The alcohol percentage of the whisky, datatype: int

**Summary of Whiskey Production and Quality by Region**

The provided data reveals interesting trends across three distinct regions: "HIGH," "LOW," and "ISLAY," concerning both the volume of whiskey produced and its average quality score.

- "HIGH" Region: This region is the dominant producer, accounting for 89 units of whiskey. However, it exhibits the lowest average quality score at approximately 74.84. This suggests that a focus on quantity might, in this instance, correlate with a slightly lower overall quality rating.
- "LOW" Region: Positioning itself as a mid-tier producer with 12 units, the "LOW" region achieves a moderate average quality score of around 77.33. Its quality is noticeably better than that of the "HIGH" region, despite significantly lower production volume.
- "ISLAY" Region: Although the smallest producer with only 8 units, the "ISLAY" region stands out for its exceptional average quality score of approximately 81.25. This indicates a strong emphasis on quality, perhaps at the expense of mass production.

The data strongly suggests an inverse relationship between production volume and average quality across these regions. While the "HIGH" region prioritizes scale, producing the most whiskey, it yields the lowest average quality. Conversely, the "ISLAY" region, with its limited output, boasts the highest average quality. The "LOW" region provides a middle ground, balancing a modest production volume with a respectable quality score that surpasses the high-volume producer. This pattern implies that smaller, more specialized production may be linked to higher perceived quality in the whiskey market, at least based on this dataset.

**Summary of Whiskey Production and Quality by District**

Based on the provided data, the highest quality whiskey in each region is produced in the following districts:

- HIGH Region: The highest average quality whiskey (90.00) is produced in the SKYE district.
- ISLAY Region: The highest average quality whiskey (84.75) is produced in the SOUTH district.
- LOW Region: The highest average quality whiskey (87.66) is produced in the CAMPBEL district.


#### Correlation between Age and Score of Whiskey

A correlation value close to zero indicates a very weak or negligible linear relationship between the two variables. This means that:

- Age has virtually no predictive power over the score of a whiskey.
- An increase or decrease in the age of whiskey does not consistently correspond to an increase or decrease in its score.
- The age of a whiskey has little to no impact on its quality score based on this dataset.







### 2. I am a fan of Bunnahabhain whisky from Islay Island. Could you recommend another whisky I might enjoy? Please explain the reasoning behind your recommendation and the factors that most influence the results.

In [38]:
df = pd.read_excel("Scotch_Whiskies_Data.xlsx", header=[0, 1])

df.columns = ['_'.join(col).strip() if isinstance(col, tuple) else col for col in df.columns.values]
df['FIN_REGION'] = df['FIN_REGION'].str.upper().str.strip()
df['FIN_DISTRICT'] = df['FIN_DISTRICT'].str.upper().str.strip()
df.rename(columns={'Unnamed: 0_level_0_NAME': 'Whisky', 'FIN_AGE': 'Age', 'FIN_SCORE': 'Score', 'FIN_%': '%', 'FIN_REGION': 'Region'}, inplace=True)


sensory_cols = [col for col in df.columns if col.startswith(('color_', 'nose_', 'body_', 'pal_', 'fin_')) and col not in ['FIN_AGE', 'FIN_SCORE', 'FIN_%']]
features_for_knn = ['%', 'Score', 'Region', 'FIN_DISTRICT'] + sensory_cols
num_features_to_scale = ['%', 'Score']


df_encoded = pd.get_dummies(df, columns=['Region', 'FIN_DISTRICT'])
df_encoded[num_features_to_scale] = MinMaxScaler().fit_transform(df_encoded[num_features_to_scale])

bunnahabhain_row = df_encoded[df_encoded['Whisky'] == 'Bunnahabhain']
bunnahabhain_row = bunnahabhain_row.iloc[0]


feature_cols = num_features_to_scale + [col for col in df_encoded.columns if 'Region_' in col] +  [col for col in df_encoded.columns if 'FIN_DISTRICT_' in col] + sensory_cols

feature_cols = [col for col in feature_cols if col in df_encoded.columns]

X = df_encoded[feature_cols]


knn = NearestNeighbors(n_neighbors=6, metric='euclidean')
knn.fit(X)


target_vector = bunnahabhain_row[feature_cols].values.reshape(1, -1)
distances, indices = knn.kneighbors(target_vector)

recommended = df.iloc[indices[0][1:]]
recommended

/Users/mikolajbankowski/Desktop/Data-Analyst-Task/venv-jp/lib/python3.10/site-packages/openpyxl/worksheet/header_footer.py:48: UserWarning: Cannot parse header or footer so it will be ignored
  warn("""Cannot parse header or footer so it will be ignored""")
/Users/mikolajbankowski/Desktop/Data-Analyst-Task/venv-jp/lib/python3.10/site-packages/openpyxl/worksheet/_reader.py:329: UserWarning: Unknown extension is not supported and will be removed
  warn(msg)
/Users/mikolajbankowski/Desktop/Data-Analyst-Task/venv-jp/lib/python3.10/site-packages/sklearn/utils/validation.py:2749: UserWarning: X does not have valid feature names, but NearestNeighbors was fitted with feature names
  warnings.warn(


Whisky  color_wyne  color_yellow  color_v.pale  color_pale  \
20       Caol Ila           1             0             0           0   
91       Pulteney           0             0             0           0   
18  Bruichladdich           0             0             0           1   
48  Glenglassaugh           0             0             0           0   
6        Balblair           1             0             0           0   

    color_p.gold  color_gold  color_o.gold  color_f.gold  color_bronze  ...  \
20             0           0             0             0             0  ...   
91             0           0             0             0             0  ...   
18             0           0             0             0             0  ...   
48             0           1             0             0             0  ...   
6              0           0             0             0             0  ...   

    FIN_arome  FIN_ling  FIN_long  FIN_very  FIN_quick  Age  Score     %  \
20          0         0         0         0          0   12     77  40.0   
91          0         0         1         0          0    8     77  40.0   
18          0         0         0         0          0   10     76  40.0   
48          0         0         0         0          0    9     76  40.0   
6           0         1         0         0          0   10     76  40.0   

    Region  FIN_DISTRICT  
20   ISLAY         NORTH  
91    HIGH         NORTH  
18   ISLAY          LOCH  
48    HIGH          SPEY  
6     HIGH         NORTH  

[5 rows x 74 columns]

The recommendations are based on the similarity between Bunnahabhain and other whiskies across several key attributes: sensory profile, region, district, %, score. The recommendation is generated using a K-Nearest Neighbors (KNN) algorithm, which identifies whiskies with the most similar characteristics to Bunnahabhain. Age is not considered in this recommendation, as indicated by the correlation analysis, which shows a negligible relationship between age and score.

The recommended whiskies are those that are most similar to Bunnahabhain.  The top recommendation, Caol Ila, is also from Islay (LOCH district), reinforcing the strong influence of geographical origin when combined with other sensory attributes. Bruichladdich is also from Islay (SOUTH district), showing a strong regional influence in the recommendations


Answer: Whisky Caol Ila
